dns-policy-operator/src/charm.py

```diff
@@ -0,0 +1,204 @@
+                    [e.strip() for e in str(self.config["allowed-hosts"]).split(",")]
+                ),
+            }
+        )
+
+    def _on_start(self, _: ops.StartEvent) -> None:
+        """Handle start event."""
+
+    def _on_install(self, _: ops.InstallEvent) -> None:
+        """Handle install event."""
+        self.unit.status = ops.MaintenanceStatus("Preparing dns-policy-app")
+        self.dns_policy.setup(self.unit.name)
+        self._timer.start(self.unit.name, "reconcile", "30s", "1m")
+
+    def _on_database_created(self, _: DatabaseCreatedEvent) -> None:
+        """Handle database created.
+
+        Args:
+            event: Event triggering the database created handler.
+        """
+        database_relation_data = self._database.get_relation_data()
+        self.unit.status = ops.MaintenanceStatus("Preparing database")
+        self.dns_policy.configure(
+            {
+                "debug": "true" if self.config["debug"] else "false",
+                "allowed-hosts": json.dumps(
+                    [e.strip() for e in str(self.config["allowed-hosts"]).split(",")]
+                ),
+                "database-port": database_relation_data["POSTGRES_PORT"],
+                "database-host": database_relation_data["POSTGRES_HOST"],
+                "database-name": database_relation_data["POSTGRES_DB"],
+                "database-password": database_relation_data["POSTGRES_PASSWORD"],
+                "database-user": database_relation_data["POSTGRES_USER"],
+            }
+        )
+        self.dns_policy.command("migrate")
+
+    def _on_database_endpoints_changed(self, _: DatabaseEndpointsChangedEvent) -> None:
+        """Handle endpoints change.
+
+        Args:
+            event: Event triggering the endpoints changed handler.
+        """
+        database_relation_data = self._database.get_relation_data()
+        self.unit.status = ops.MaintenanceStatus("Preparing database")
+        self.dns_policy.configure(
+            {
+                "debug": "true" if self.config["debug"] else "false",
+                "allowed-hosts": json.dumps(
+                    [e.strip() for e in str(self.config["allowed-hosts"]).split(",")]
+                ),
+                "database-port": database_relation_data["POSTGRES_PORT"],
+                "database-host": database_relation_data["POSTGRES_HOST"],
+                "database-name": database_relation_data["POSTGRES_DB"],
+                "database-password": database_relation_data["POSTGRES_PASSWORD"],
+                "database-user": database_relation_data["POSTGRES_USER"],
+            }
+        )
+        self.dns_policy.command("migrate")
+
+    def _on_create_reviewer_action(self, event: ops.charm.ActionEvent) -> None:
+        """Handle the create reviewer ActionEvent.
+
+        Args:
+            event: Event triggering this action handler.
+        """
+        event.set_results(
+            {
+                "result": self.dns_policy.command(
```

Comment:
You should probably catch a `CommandError` when executing the cmd and set the result accordingly, failing with a friendly error message and logging the exception if something occurs

Reply:
I improved the logging and the user facing message in case of failure